File: music-commons-core/src/main/java/net/transgressoft/commons/music/playlist/MutablePlaylistDirectory.java
package net.transgressoft.commons.music.playlist;

import com.google.common.base.MoreObjects;
import com.google.common.base.Objects;
import net.transgressoft.commons.music.audio.AudioItem;

import java.util.Collections;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ConcurrentSkipListSet;

import static java.util.Objects.requireNonNull;

class MutablePlaylistDirectory<I extends AudioItem> extends MutablePlaylist<I> implements MutableAudioPlaylistDirectory<I> {

    private final Set<AudioPlaylist<I>> descendantPlaylists;

    protected MutablePlaylistDirectory(int id, String name) {
        this(id, name, Collections.emptyList());
    }

    protected MutablePlaylistDirectory(int id, String name, List<I> audioItems) {
        this(id, name, audioItems, Collections.emptySet());
    }

    protected <N extends AudioPlaylist<I>> MutablePlaylistDirectory(int id, String name, List<I> audioItems, Set<N> playlists) {
        super(id, name, audioItems);
        descendantPlaylists= new ConcurrentSkipListSet<>(playlists);
    }

    @Override
    @SafeVarargs
    public final <N extends AudioPlaylist<I>> void addPlaylist(N... playlists) {
        requireNonNull(playlists);
        addAllPlaylists(Set.of(playlists));
    }

    @Override
    public <N extends AudioPlaylist<I>> void addAllPlaylists(Set<N> playlists) {
        requireNonNull(playlists);
        descendantPlaylists.addAll(playlists);
    }

    @Override
    public <N extends AudioPlaylist<I>> void removePlaylist(N playlist) {
        descendantPlaylists.removeIf(p -> p.equals(playlist));
    }

    @Override
    public <N extends AudioPlaylist<I>> boolean containsPlaylist(N playlist) {
        return descendantPlaylists.contains(playlist);
    }

    @Override
    public <N extends AudioPlaylist<I>> Set<N> descendantPlaylists() {
        return (Set<N>) descendantPlaylists;
    }

    @Override
    public boolean isDirectory() {
        return true;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        var that = (MutablePlaylistDirectory<I>) o;
        return Objects.equal(getName(), that.getName()) && Objects.equal(id(), that.id());
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(getName(), id());
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("id", id())
                .add("name", getName())
                .add("descendantPlaylists", descendantPlaylists.size())
                .add("audioItems", audioItems().size())
                .toString();
    }
}

File: music-commons-core/src/main/kotlin/net/transgressoft/commons/music/playlist/MutablePlaylistDirectory.kt
package net.transgressoft.commons.music.playlist

import com.google.common.base.Objects
import net.transgressoft.commons.music.audio.AudioItem

internal open class MutablePlaylistDirectory<I : AudioItem, N : AudioPlaylist<I>>(
    id: Int,
    theName: String,
    audioItems: List<I>? = null,
    playlists: Set<N>? = null,
) : ImmutablePlaylistDirectory<I, N>(id, theName, audioItems, playlists), MutableAudioPlaylistDirectory<I, N> {

    override var name: String = theName
        set(value) {
            super.setNameInternal(value)
        }

    override fun addAudioItems(audioItems: Collection<I>) {
        super.addAll(audioItems)
    }

    override fun removeAudioItems(audioItems: Collection<I>) {
        super.removeAll(audioItems)
    }

    override fun clearAudioItems() {
        super.clear()
    }

    override fun addPlaylists(playlists: Set<N>) {
        super.addAll(playlists)
    }

    override fun removePlaylists(playlists: Set<N>) {
        super.removeAll(playlists)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as MutablePlaylistDirectory<*, *>
        return Objects.equal(name, that.name) && Objects.equal(id, that.id)
    }

    override fun hashCode() = Objects.hashCode(name, id)
}